internal/manifest/docker_schema2_list.go

```diff
@@ -69,23 +78,55 @@ func (list *Schema2ListPublic) Instance(instanceDigest digest.Digest) (ListUpdat
-			return fmt.Errorf("update %d of %d passed to Schema2List.UpdateInstances contained an invalid digest: %w", i+1, len(updates), err)
-		}
-		list.Manifests[i].Digest = updates[i].Digest
-		if updates[i].Size < 0 {
-			return fmt.Errorf("update %d of %d passed to Schema2List.UpdateInstances had an invalid size (%d)", i+1, len(updates), updates[i].Size)
-		}
-		list.Manifests[i].Size = updates[i].Size
-		if updates[i].MediaType == "" {
-			return fmt.Errorf("update %d of %d passed to Schema2List.UpdateInstances had no media type (was %q)", i+1, len(updates), list.Manifests[i].MediaType)
+	return index.EditInstances(editInstances)
+}
+
+func (index *Schema2ListPublic) EditInstances(editInstances []ListEdit) error {
+	addedEntries := []Schema2ManifestDescriptor{}
+	for i, editInstance := range editInstances {
+		switch editInstance.ListOperation {
+		case listOpUpdate:
+			if err := editInstance.UpdateOldDigest.Validate(); err != nil {
+				return fmt.Errorf("Schema2List.EditInstances: Attempting to update %s which is an invalid digest: %w", editInstance.UpdateOldDigest, err)
+			}
+			if err := editInstance.UpdateDigest.Validate(); err != nil {
+				return fmt.Errorf("Schema2List.EditInstances: Modified digest %s is an invalid digest: %w", editInstance.UpdateDigest, err)
+			}
+			instanceToBeModifiedIdx := index.descriptorIndex(editInstance.UpdateOldDigest)
+			if instanceToBeModifiedIdx < 0 {
+				return fmt.Errorf("Schema2List.EditInstances: Attempting to update %s which is an invalid digest", editInstance.UpdateOldDigest)
+			}
+			index.Manifests[instanceToBeModifiedIdx].Digest = editInstance.UpdateDigest
+			if editInstance.UpdateSize < 0 {
+				return fmt.Errorf("update %d of %d passed to Schema2List.UpdateInstances had an invalid size (%d)", i+1, len(editInstances), editInstance.UpdateSize)
+			}
+			index.Manifests[instanceToBeModifiedIdx].Size = editInstance.UpdateSize
+			if editInstance.UpdateMediaType == "" {
+				return fmt.Errorf("update %d of %d passed to Schema2List.UpdateInstances had no media type (was %q)", i+1, len(editInstances), index.Manifests[i].MediaType)
+			}
+			index.Manifests[instanceToBeModifiedIdx].MediaType = editInstance.UpdateMediaType
+		case listOpAdd:
+			addInstance := Schema2ManifestDescriptor{
+				Schema2Descriptor{Digest: editInstance.AddDigest, Size: editInstance.AddSize, MediaType: editInstance.AddMediaType},
+				Schema2PlatformSpec{
+					OS:           editInstance.AddPlatform.OS,
+					Architecture: editInstance.AddPlatform.Architecture,
+					OSVersion:    editInstance.AddPlatform.OSVersion,
+					OSFeatures:   editInstance.AddPlatform.OSFeatures,
+					Variant:      editInstance.AddPlatform.Variant,
+				},
+			}
+			addedEntries = append(addedEntries, addInstance)
+		default:
+			return fmt.Errorf("internal error: invalid operation: %d", editInstance.ListOperation)
 		}
-		list.Manifests[i].MediaType = updates[i].MediaType
+		index.Manifests = append(index.Manifests, addedEntries...)
```

Comment:
(I’d prefer the handling of `addedEntries` here to directly correspond to the OCI handling, whatever that will be.)

Reply:
We don't add zstd annotations to docker schema right, hence not sorting here.

Reply:
You’re right, the two would differ in sorting.

But the rest of the logic can directly correspond — see how the schema2 variant now turned out to be incorrect.

Having the two implementations very similar decreases the risk of such problems, and generally allows future maintenance to be done with half the analysis.

Reply:
Fixed.